### PR TITLE
Order the acquisitions per modality based on series number: Redmine 10959

### PR DESCRIPTION
**Pull request by MounaSafiHarab**
Order the files with the same AcquisitionProtocolID (within a tarchiveID) by series_number.
https://github.com/aces/Loris/pull/2221

### CODE REVIEW

**Comment by cmadjar** [comment]
Is there a script or patch that can be run on files that were already inserted into the DB?

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 1]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+
+    my $this = shift;
+    my ($row, $file, $acqProtID, $dataArr, $cmd, $message, $sth);
+    my ($tarchiveID, $tarchive_srcloc)= @_;
+    my $upload_id = getUploadIDUsingTarchiveSrcLoc($tarchive_srcloc);
+
+    my $query = "SELECT DISTINCT f.AcquisitionProtocolID ".
+		"from files f ";
+    my $where = "WHERE f.TarchiveSource=?";   
+    $query = $query . $where;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: The $where variable is not useful. I would write the whole query directly in the $query variable.
Also, FROM was not capitalized here (not crucial but to stay consistent)

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 2]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+    my $acqArr = ${$this->{'dbhr'}}->prepare($query);
+    $acqArr->execute($tarchiveID);
+    # For each of the files having this AcquisitionProtocolID
+    # load the file object to get the series_number  
+    while (my $row = $acqArr->fetchrow_hashref()) {
+	$acqProtID = $row->{'AcquisitionProtocolID'};
+	$query = "SELECT f.FileID, f.ModalityOrder ".
+		 "from files f ";
+	$where = "WHERE f.TarchiveSource=? AND f.AcquisitionProtocolID=?";
+     	$query = $query . $where;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Same here. Get rid of the $where variable and capitalize "FROM"

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 3]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+
+	if ($this->{debug}) {
+	    print $query . "\n";
+	}
+
+	$dataArr = ${$this->{'dbhr'}}->prepare($query);
+	$dataArr->execute($tarchiveID, $acqProtID);
+	my (@fileID, @seriesNumber)=();
+	my ($fileID, $seriesNumber)=();
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Why is this line repeated with @ (array) and $ (dereferenced array)?

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 4]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+	my $i=0;
+    	while (my $row2 = $dataArr->fetchrow_hashref()) {
+	    $i++;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Incrementation should be put at the end of the loop as arrays start at index 0.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 5]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+	    $seriesNumber[$i] = $file->getParameter('series_number');
+	}
+    	my (@sorted_seriesNumber_indices, @sorted_fileID)=();
+	# Sort the series_number, and assign the Modality Order accordingly
+    	@sorted_seriesNumber_indices = sort {$seriesNumber[$a] <=> $seriesNumber[$b]} (0..$#seriesNumber);
+	@sorted_fileID = @fileID[@sorted_seriesNumber_indices];
+
+	my $order = 0;
+	foreach my $j (0..$#seriesNumber-1) {
+	     $order++;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Will need to move incrementation at the end too.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 6]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+            $query = "UPDATE files f SET f.ModalityOrder=? ";
+	     $where = "WHERE f.FileID=?";
+	     $query = $query . $where;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Same comment as before, the $where variable is not useful here. Remove it and write the query in one shot.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 7]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+            if ($this->{debug}) {
+                print $query . "\n";
+	     }
+            my $modalityOrder_update = ${$this->{'dbhr'}}->prepare($query);
+            $modalityOrder_update->execute($order, $sorted_fileID[$order]);
+            $message = "The Modality Order for FileID $sorted_fileID[$order] was updated to $order \n ";
+	    print "Message is: $message \n";
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: get rid of the print statement here and keep it only for the LOGs

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (hunk 1 above)
Comment (MounaSafiHarab) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: done.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 8]
```diff
@@ -1242,6 +1242,80 @@ sub computeSNR {
...
+        print $query . "\n";
+    }
+
+    my $acqArr = ${$this->{'dbhr'}}->prepare($query);
+    $acqArr->execute($tarchiveID);
+    # For each of the files having this AcquisitionProtocolID
+    # load the file object to get the series_number  
+    while (my $row = $acqArr->fetchrow_hashref()) {
+	$acqProtID = $row->{'AcquisitionProtocolID'};
+	$query = "SELECT f.FileID, f.ModalityOrder ".
```
Comment (MounaSafiHarab) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: no, you understood it right. I misnamed it here (it was the original name I chose but then I updated it in Loris but forgot to update here). Thanks for the catch!
done now

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (hunk 2 above)
Comment (MounaSafiHarab) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: done

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (hunk 6 above)
Comment (MounaSafiHarab) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: done

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (hunk 7 above)
Comment (MounaSafiHarab) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: done

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 9]
```diff
@@ -1242,6 +1242,75 @@ sub computeSNR {
...
+	     $order++;
+            $query = "UPDATE files f SET f.AcqOrderPerModality=? ".
+	              "WHERE f.FileID=?";
+            if ($this->{debug}) {
+                print $query . "\n";
+	     }
+            my $modalityOrder_update = ${$this->{'dbhr'}}->prepare($query);
+            $modalityOrder_update->execute($order, $sorted_fileID[$order]);
+            $message = "The Modality Order for FileID $sorted_fileID[$order] was updated to $order \n ";
+	    $this->{LOG}->print($message);
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: alignment issue

**Comment by MounaSafiHarab** [comment]
@cmadjar 
I took care of the rest of the comments. It should be ready if you want to test it again.

re. the comment on populating existing data; I will need to generate a script. I am not sure this is doable through a simple mysql patch.

**Comment by cmadjar** [comment]
Works like a charm :) 
screen shot 2016-10-13 at 10 39 32 am

**Pull request merged**